src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+  b.buckets.forM (AssocList.forM f)
+
+/-- Support for the `for` loop construct in `do` blocks. -/
+@[inline] def forIn (f : (a : α) → β a → δ → m (ForInStep δ)) (init : δ) (b : Raw α β) : m δ :=
+  b.buckets.forIn init (fun bucket acc => bucket.forInStep acc f)
+
+instance : ForM m (Raw α β) ((a : α) × β a) where
+  forM m f := m.forM (fun a b => f ⟨a, b⟩)
+
+instance : ForIn m (Raw α β) ((a : α) × β a) where
+  forIn m init f := m.forIn (fun a b acc => f ⟨a, b⟩ acc) init
+
+/-- Transforms the hash map into a list of mappings in some order. -/
+@[inline] def toList (m : Raw α β) : List ((a : α) × β a) :=
+  m.foldl (fun acc k v => ⟨k, v⟩ :: acc) []
+
+/-- Transforms the hash map into an array of mappings in some order. -/
+@[inline] def toArray (m : Raw α β) : Array ((a : α) × β a) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+@[inline, inherit_doc Raw.toList] def Const.toList {β : Type v} (m : Raw α (fun _ => β)) :
+    List (α × β) :=
+  m.foldl (fun acc k v => ⟨k, v⟩ :: acc) []
+
+@[inline, inherit_doc Raw.toArray] def Const.toArray {β : Type v} (m : Raw α (fun _ => β)) :
+    Array (α × β) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+/-- Returns a list of all keys present in the hash map in some order. -/
+@[inline] def keys (m : Raw α β) : List α :=
+  m.foldl (fun acc k _ => k :: acc) []
+
+/-- Returns an array of all keys present in the hash map in some order. -/
+@[inline] def keysArray (m : Raw α β) : Array α :=
+  m.foldl (fun acc k _ => acc.push k) #[]
+
+/-- Returns a list of all values present in the hash map in some order. -/
+@[inline] def values {β : Type v} (m : Raw α (fun _ => β)) : List β :=
+  m.foldl (fun acc _ v => v :: acc) []
+
+/-- Inserts multiple mappings into the hash map by iterating over the given collection and calling
+`insert`. If the same key appears multiple times, the last occurrence takes precendence. -/
+@[inline] def insertMany [BEq α] [Hashable α] {ρ : Type w} [ForIn Id ρ ((a : α) × β a)]
+    (m : Raw α β) (l : ρ) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.insertMany ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+@[inline, inherit_doc Raw.insertMany] def Const.insertMany {β : Type v} [BEq α] [Hashable α]
+    {ρ : Type w} [ForIn Id ρ (α × β)] (m : Raw α (fun _ => β)) (l : ρ) : Raw α (fun _ => β) :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.Const.insertMany ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+@[inline, inherit_doc Raw.insertMany] def Const.insertManyUnit [BEq α] [Hashable α] {ρ : Type w}
+    [ForIn Id ρ α] (m : Raw α (fun _ => Unit)) (l : ρ) : Raw α (fun _ => Unit) :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.Const.insertManyUnit ⟨m, h⟩ l).1
+  else m -- will never happen for well-formed inputs
+
+/-- Creates a hash map from a list of mappings. -/
+@[inline] def ofList [BEq α] [Hashable α] {ρ : Type w} [ForIn Id ρ ((a : α) × β a)] (l : ρ) :
```

Comment:
The docstring and name don't seem to match the type signature here. Should the type be specialized?

Reply:
We definitely want a function with that name for recognizability, even though it is just a special case of `insertMany`. So the remaining question is: if the type signature of the function can be generalized without any disadvantages, is there a reason not to do it? If I actually call the function with a list, then the compiler specializes the IR to be just as efficient as if the function had been declared for lists from the start. The only reason I see why you would want to explicitly specialize to lists here is that Loogle might find the function on a query like `List, Std.HashMap`?

Reply:
Search is important. The other reason to want the monomorphised version is that it can help type inference and improve error messages.

If I pass a value to `ofList`, I'm presumably doing it because I believe it to be a list; the monomorphised function shares this belief with Lean. This could solve metavariables elsewhere in my program and make it typecheck, for example.